pdns/dnsdist.cc

```diff
@@ -577,146 +578,111 @@ void handleResponseSent(const IDState& ids, double udiff, const ComboAddress& cl
   }
 }
 
+void processResponderPacket(std::shared_ptr<DownstreamState>& dss, PacketBuffer& response, LocalStateHolder<std::vector<DNSDistResponseRuleAction>>& localRespRuleActions, IDState* ids, bool& shouldRelease, int& delayMsec);
 // listens on a dedicated socket, lobs answers from downstream servers to original requestors
 void responderThread(std::shared_ptr<DownstreamState> dss)
 {
   try {
-  setThreadName("dnsdist/respond");
-  auto localRespRuleActions = g_respruleactions.getLocal();
-  const size_t initialBufferSize = getInitialUDPPacketBufferSize();
-  PacketBuffer response(initialBufferSize);
-
-  /* when the answer is encrypted in place, we need to get a copy
-     of the original header before encryption to fill the ring buffer */
-  dnsheader cleartextDH;
-  uint16_t queryId = 0;
-  std::vector<int> sockets;
-  sockets.reserve(dss->sockets.size());
-
-  for(;;) {
-    try {
-      dss->pickSocketsReadyForReceiving(sockets);
-      if (dss->isStopped()) {
-        break;
-      }
-
-      for (const auto& fd : sockets) {
-        response.resize(initialBufferSize);
-        ssize_t got = recv(fd, response.data(), response.size(), 0);
-
-        if (got == 0 && dss->isStopped()) {
-          break;
-        }
-
-        if (got < 0 || static_cast<size_t>(got) < sizeof(dnsheader)) {
-          continue;
-        }
-
-        response.resize(static_cast<size_t>(got));
-        dnsheader* dh = reinterpret_cast<struct dnsheader*>(response.data());
-        queryId = dh->id;
-
-        IDState* ids = dss->getExistingState(queryId);
-        if (ids == nullptr) {
-          continue;
-        }
-
-        int64_t usageIndicator = ids->usageIndicator;
-
-        if (!IDState::isInUse(usageIndicator)) {
-          /* the corresponding state is marked as not in use, meaning that:
-             - it was already cleaned up by another thread and the state is gone ;
-             - we already got a response for this query and this one is a duplicate.
-             Either way, we don't touch it.
-          */
-          continue;
+    setThreadName("dnsdist/respond");
+    auto localRespRuleActions = g_respruleactions.getLocal();
+
+#ifdef HAVE_XSK
+    if (dss->xskInfo) {
+      dss->setUp();
```

Comment:
Why is `setUp()` called here? Are the regular health-check mechanisms not working in that case?

Reply:
Oh yes.  Currently AF_XDP is wrong to work with health checks.  Because AF_XDP will intercept all packets and make the health check timeout.  I called setup() to avoid the trouble of health checks when debugging.  I forgot that this place wasn't handled properly.  I will fix it today.

Reply:
Now xsk works properly with healthcheck.